xnmt/modelparts/scorers.py

```diff
@@ -126,27 +127,29 @@ def __init__(self,
   
   def calc_scores(self, x: dy.Expression) -> dy.Expression:
     return self.output_projector.transform(x)
+  
+  def is_modifying_softmax_layer(self):
```

Comment:
So basically it means that we can't use dy.pickneglogsoftmax because we are modifying the value of the softmax layer before picking the loss. The dy.pickneglogsoftmax is a convenient function but can't be used when we are doing: 
1. label smoothing
2. Linear interpolation on softmax

So this method is called to check that value, and use the correct dynet function accordingly. Also, by this inheritance, I don't have to copy+paste the calc_loss function from the Softmax object.

Reply:
Ok, let me check again. I think "modifying" is too vague though. Try to use a more descriptive name.

Reply:
Ok, let me check again. I think "modifying" is too vague though. Try to use a more descriptive name.

Reply:
Can you tell me one good name for this operation?

Reply:
OK, let me confirm: what do you mean by "pickneglogsoftmax cannot be used"? Why can it not be used? To elaborate,

* `calc_scores()` calculates an unnormalized log probability
* `calc_log_probs()` calculates a normalized log probability
* `calc_probs()` calculates a normalized probability

As long as these functions are implemented correctly, we should always be able to use pickneglogsoftmax, correct? If any of these functions don't do the desired action in your implementation here, we should fix the functions. This is an absolute must.

Or, if these are an efficiency concern where in some cases it's faster/memory efficient to calculate scores (log probs), and in some cases it's faster/memory efficient to calculate probabilities, we could have a function called something like `prefer_scores_to_probs()`, where if it's true we use the `calc_scores()` and if it's false we use `calc_probs()`.

Does this make sense?

Reply:
Yep these are correct:
- calc_scores() calculates an unnormalized log probability
- calc_log_probs() calculates a normalized log probability
- calc_probs() calculates a normalized probability

```
As long as these functions are implemented correctly, we should always be able to use pickneglogsoftmax, correct?
```
I think this is not correct. For example the current implementation of master branch separates when to use ```dy.pick``` and ```dy.pickneglogsoftmax``` 

https://github.com/neulab/xnmt/blob/c6446608d94aabeda3c2dca7bbd2920f23f593f7/xnmt/modelparts/scorers.py#L134-L149

This is because the loss itself can't be derrived by looking at the score of the unnormalized log prob of the model. For example, in the case when we are interpolating the softmax probability directly the loss should be the log probability of the interpolated model, right?